tests/common/include/tests/common/scattering.inl

```diff
@@ -0,0 +1,128 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Project include(s).
+#include "detray/propagator/actors/pointwise_material_interactor.hpp"
+#include "detray/simulation/detail/scattering_helper.hpp"
+#include "detray/simulation/random_scatterer.hpp"
+#include "detray/tracks/tracks.hpp"
+#include "detray/utils/statistics.hpp"
+
+// google-test include(s).
+#include <gtest/gtest.h>
+
+// System include(s).
+#include <random>
+#include <vector>
+
+using namespace detray;
+using transform3 = __plugin::transform3<detray::scalar>;
+using matrix_operator = typename transform3::matrix_actor;
+using vector3 = typename transform3::vector3;
+
+// Test scattering helper
+TEST(scattering, scattering_helper) {
+
+    std::random_device rd{};
+    std::mt19937_64 generator{rd()};
+    generator.seed(0u);
+
+    const vector3 dir{2.f, 2.f, 4.f};
+    const scalar scattering_angle{0.1f};
+
+    // xaxis of curvilinear plane
+    const vector3 u = unit_vectors<vector3>().make_curvilinear_unit_u(dir);
+
+    // normal vector of the plane defined by u and dir
+    const vector3 w = vector::cross(dir, u);
+
+    std::vector<scalar> angle_diffs;
+
+    std::size_t n_samples{100000u};
+    for (std::size_t i = 0u; i < n_samples; i++) {
+
+        // Scatter the input direction
+        const vector3 new_dir =
+            scattering_helper<transform3>()(dir, scattering_angle, generator);
+
+        // Assign a plus sign if the dot product of new direction and w vector
+        // is plus
+        const scalar sign = vector::dot(w, new_dir) > 0 ? 1.f : -1.f;
+
+        // the cosine of angle between original direction and new one
+        scalar cos_theta = vector::dot(dir, new_dir) /
+                           (getter::norm(dir) * getter::norm(new_dir));
+
+        // To prevent rounding error where cos_theta is out of [-1, 1]
+        cos_theta = std::min(cos_theta, scalar{1.f});
```

Comment:
How about ```std::clamp```?

Reply:
I think I will try to get the angle difference using cross product rather than numerically instable dot product

Reply:
Just from intuition, but might the cross product not rather suffer from catastrophic cancellation?

Reply:
In the cross product you take a lot of differences and if a difference is close to zero, the result can have a very bad error coming from the floating point approximation. So, when the vectors are almost parallel, the cross product could yield a very bad result...

Reply:
I haven't checked the numbers, so maybe this won't actually happen, idk

Reply:
It depends on the size of error. At least I can tolerate small errors; asin() won't spit NaN value as long as the cross product doesn't spike to -10 or so

Reply:
Just saying, I suspect (not sure) the relative error stemming from floating point calculations might give you a more noisy approximation to the scattering angle if calculated with the cross product. If you can deal with that error better this way, Ok

Reply:
So, not using ```std::clamp``` instead of ```min``` and ```max```?